astropy/coordinates/transformations.py

```diff
@@ -746,8 +749,157 @@ def __call__(self, fromcoord, toframe):
+    finite_difference_dt : Quantity or callable
+        If a quantity, this is the size of the differential used to do the
+        finite difference.  If a callable, should accept
+        ``(fromcoord, toframe)`` and return the ``dt`` value.
+    symmetric_finite_difference : bool
+        If True, the finite difference is computed as
+        :math:`\frac{x(t + \Delta t / 2) - x(t + \Delta t / 2)}{\Delta t}`, or
+        if False, :math:`\frac{x(t + \Delta t) - x(t)}{\Delta t}`.  The latter
+        case has slightly better performance (and more stable finite difference
+        behavior).
+
+    All other paramters are identical to the initializer for
+    `FunctionTransform`.
+
+    """
+    def __init__(self, func, fromsys, tosys, priority=1, register_graph=None,
+                 finite_difference_frameattr_name='obstime',
+                 finite_difference_dt=1*u.second,
+                 symmetric_finite_difference=True):
+        super(FunctionTransformWithFiniteDifference, self).__init__(func,
+              fromsys, tosys, priority, register_graph)
+        self.finite_difference_frameattr_name = finite_difference_frameattr_name
+        self.finite_difference_dt = finite_difference_dt
+        self.symmetric_finite_difference = symmetric_finite_difference
+
+    @property
+    def finite_difference_frameattr_name(self):
+        return self._finite_difference_frameattr_name
+    @finite_difference_frameattr_name.setter
+    def finite_difference_frameattr_name(self, value):
+        if value is None:
+            self._diff_attr_in_fromsys = self._diff_attr_in_tosys = False
+        else:
+            diff_attr_in_fromsys = value in self.fromsys.frame_attributes
+            diff_attr_in_tosys = value in self.tosys.frame_attributes
+            if diff_attr_in_fromsys or diff_attr_in_tosys:
+                self._diff_attr_in_fromsys = diff_attr_in_fromsys
+                self._diff_attr_in_tosys = diff_attr_in_tosys
+            else:
+                raise ValueError('Frame attribute name {} is not a frame '
+                                 'attribute of {} or {}'.format(value,
+                                                                self.fromsys,
+                                                                self.tosys))
+        self._finite_difference_frameattr_name = value
+
+    def __call__(self, fromcoord, toframe):
+        from .representation import (CartesianRepresentation,
+                                     CartesianDifferential)
+
+        supcall = self.func
+        if getattr(fromcoord.data, 'differentials', False):
+            # this is the finite difference case
+
+            if callable(self.finite_difference_dt):
+                dt = self.finite_difference_dt(fromcoord, toframe)
+            else:
+                dt = self.finite_difference_dt
+
+            fromdiffless = fromcoord.realize_frame(fromcoord.data.without_differentials())
```

Comment:
much Nitpick: inconsistent use of underscores in variable names here and below...